File: 00452-minimum-number-of-arrows-to-burst-balloons/main.js
/**
 * @param {number[][]} points
 * @return {number}
 */
var findMinArrowShots = function (points) {
  const ranges = points.map(([x0, x1]) => ({ start: x0, end: x1 }));
  return findMaxInter(ranges);
};
function isInRange(x, ran) {
  let x0 = ran.start;
  let x1 = ran.end;
  return x0 <= x && x <= x1;
}

function findMaxInter(points) {
  let ans = 1;
  points.sort((a, b) => a.start - b.start);
  console.log(points);
  let currentRange = points[0];
  for (let i = 1; i < points.length; i++) {
    let x0 = points[i].start;
    let x1 = points[i].end;
    if (isInRange(x0, currentRange)) {
      currentRange = {
        start: Math.max(currentRange.start, x0),
        end: Math.min(currentRange.end, x1),
      };
    } else {
      currentRange = { start: x0, end: x1 };
      ans++;
    }
    console.log("current range", currentRange);
  }
  return ans;
}

const points = [
  [3, 9],
  [7, 12],
  [3, 8],
  [6, 8],
  [9, 10],
  [2, 9],
  [0, 9],
  [3, 9],
  [0, 6],
  [2, 8],
];
const ans = findMaxInter(points);
console.log("ans", ans);

File: 00452-minimum-number-of-arrows-to-burst-balloons/main.py
from typing import List


def is_in_range(x, ran):
    x0, x1 = ran
    return x0 <= x <= x1


class Solution:
    def findMinArrowShots(self, points: List[List[int]]) -> int:
        ans = 1
        points.sort()
        print(points)
        current_range = points[0]
        for (x0, x1) in points[1:]:
            if is_in_range(x0, current_range):
                current_range = [max(current_range[0], x0),
                                 min(current_range[1], x1)]
            else:
                current_range = [x0, x1]
                ans += 1
            # print("current range", current_range)
        return ans


sol = Solution()
ans = sol.findMinArrowShots(
    [[3, 9], [7, 12], [3, 8], [6, 8], [9, 10], [2, 9], [0, 9], [3, 9], [0, 6], [2, 8]])
print("ans", ans)
